### DISCUSSION TITLE
UI Design on a Denny's Napkin

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#350] [Announcements]
I wanted to provide some brief direction for the UI effort, as it looks like it's caught up with the meager directions provided!

Hopefully you can make some sense from this!  

I'd like to see all effects in a vertical stack column, as I believe was the case in the most recent.

![3BE5EA99-0BCC-4D1E-8AF5-16260F772DB7_1_102_o](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/3698967/d4cac0a1-694e-4cc4-b52b-2cfbcc49fc52)

In this sketch, you'll find the following:

- The Now Playing section at the top, which includes a preview of what that matrix is currently showing.  My understanding is this might need me to convert the socket to a websocket - will need more info from whomever works on this
- App settings icon in the top right takes you to the app global settings page
- Brightness slider
- Effect settings button

These remain onscreen and do not scroll, but everything below it scrolls as a list.  On a wide display like a monitor hosting an HTML page, these should be constrained in width to something visually appealing.

Each effect has an entry here. Effects can be dragged to reorder.  The settings page for an effect (TBD) includes a Duplicate Effect and Remove Effect options.  Only copies can be removed, not the original built in effect.  The intent here is you could copy an effect twice and adjust the settings differently for all three copies.

On the far left of the effect entry is an enable/disable slider.  At least one effect must be active, so an error message should tell the user that if they try to disable the last effect.  The effect title is left justified against the slider.  The settings icon and play button are right-justified.  The currently playing effect should light up the play button, others show it in dim green.

Clicking on the effects settings icon takes you to the settings page for that effect (TBD).

**Comment by smilima** [comment]
I just bought a 64x64 HUB75 display that will provide room for the UI.  

I think the 64x32 boards should have hidden UI’s above and below the current effect (your scroll break hint, but one on top and bottom).  This will allow use of directional arrows on the larger remote. 

Now, implementing this is a different story!

**Reply by smilima** [reply]
Right after I replied to this I realized this was for the web interface…

However, I would like to see this on the matrix display!

**Comment by robertlipe** [comment]
UI in 64x64px is hard.

See my open bug report on 64line HUB75Es not working. I've been meaning to
build up a board from scratch and see about getting a logic analyzer in
that E line.

**Comment by smilima** [comment]
Interesting. Have you traced the E pin and board visually and see if it
connects to the wrover?

Thanks for the heads up.

**Comment by robertlipe** [comment]
No. Multi-layer PCB + Uncertainty what pin on WROVER _any_ of the HUB75
pins are connected to. I haven't yet found a declaration or mapping of the
HUB75 pins like we have for LED_PINn and friends. Have you sussed that out?

I know there's _something_ more required than just plugging in a tall
panel. Doubling the number of pixels we're required to shove around may be
entertaining, too.

P.S. My gmail address is the same as my name here.

**Comment by davepl** [comment]
Is this what you’re looking for?  It’s in MatrixHardware_ESP32_Custom.h as required by the SmartMatrix package.

    #define R1_PIN  GPIO_NUM_2                  
    #define G1_PIN  GPIO_NUM_0                  
    #define B1_PIN  GPIO_NUM_32                 
    #define R2_PIN  GPIO_NUM_25                 
    #define G2_PIN  GPIO_NUM_33                 
    #define B2_PIN  GPIO_NUM_27                 
    #define A_PIN   GPIO_NUM_5                  
    #define B_PIN   GPIO_NUM_4                  
    #define C_PIN   GPIO_NUM_19                 
    #define D_PIN   GPIO_NUM_18                 
    #define E_PIN   GPIO_NUM_26                 
    #define LAT_PIN GPIO_NUM_21                 
    #define OE_PIN  GPIO_NUM_23                 
    #define CLK_PIN GPIO_NUM_22

**Comment by robertlipe** [comment]
Yes, Dave. That's helpful. Thank you. My indexer hadn't picked it up as it's the only uppercase filename in the project. Oops.
Knowing that the E line IS present and that it at least used to work is quite helpful.  Now, with a bit more context and concentration, this line being NOT dependent upon MATRIX_HEIGHT in some way looks WAY suspicious:

    static const uint8_t kPanelType = SMARTMATRIX_HUB75_32ROW_MOD16SCAN;                // use SMARTMATRIX_HUB75_16ROW_MOD8SCAN for common 16x32 panels

Was that maybe ...75_64ROW_... in your earlier successful tests?

It'll probably be Tuesday before I can come back to that...
